### PR TITLE
test disabled inheritance for spaces

### PR DESCRIPTION
**Pull request by mmulholla**
replace toolchain-common for testing. https://github.com/codeready-toolchain/toolchain-common/pull/337

Also includes a new test for disabled inheritance with spaces.

### CODE REVIEW

File: test/e2e/parallel/space_test.go [hunk 1]
```diff
@@ -478,6 +478,29 @@ func TestSubSpaces(t *testing.T) {
...
+		VerifyResourcesProvisionedForSpace(t, awaitilities, parentSpace.Name, UntilSpaceHasStatusTargetCluster(memberAwait.ClusterName))
+		_, err := hostAwait.WaitForMasterUserRecord(t, parentSpaceBindings.Spec.MasterUserRecord)
+		require.NoError(t, err)
+
+		// when
+		// we also have a subSpace with same tier but with disable inheritance
+		subSpace := CreateSubSpace(t, awaitilities, testspace.WithSpecParentSpace(parentSpace.Name), testspace.WithTierName("appstudio"), testspace.WithSpecTargetCluster(memberAwait.ClusterName), testspace.WithDisableInheritance(true))
+
+		// wait until subSpace has been provisioned
+		_, subSpaceNSTemplateSet := VerifyResourcesProvisionedForSpace(t, awaitilities, subSpace.Name)
```
Comment (mfrancisc) [inline test/e2e/parallel/space_test.go:496]: this function expects some space roles to be present, while the subSpace has none if I'm not wrong. I think this is also the cause of the current test failure.

**Review by mfrancisc — COMMENTED** [review]
Looks good, nice job!

I have few general questions:
- should we create subSpace with no inheritance using SpaceRequests, so that we verify if the disableInheritance parameter is being propagated correctly from SpaceRequest to Space objects. Or this will be a separate test? 

- should we create bigger hierarchy ( 3 spaces at least )

WDYT?

File: test/e2e/parallel/space_test.go [hunk 2]
```diff
@@ -480,6 +480,39 @@ func TestSubSpaces(t *testing.T) {
...
+		parentSpace, _, parentSpaceBindings := CreateSpace(t, awaitilities, testspace.WithSpecTargetCluster(memberAwait.ClusterName), testspace.WithTierName("appstudio"))
+		// then
+		// wait until MUR and Space have been provisioned
+		VerifyResourcesProvisionedForSpace(t, awaitilities, parentSpace.Name, UntilSpaceHasStatusTargetCluster(memberAwait.ClusterName))
+		mur, err := hostAwait.WaitForMasterUserRecord(t, parentSpaceBindings.Spec.MasterUserRecord)
+		require.NoError(t, err)
+
+		// when
+		// we also have a subSpace with same tier but with disable inheritance
+		subSpace := CreateSubSpace(t, awaitilities, testspace.WithSpecParentSpace(parentSpace.Name), testspace.WithTierName("appstudio"), testspace.WithSpecTargetCluster(memberAwait.ClusterName), testspace.WithDisableInheritance(true))
```
Comment (MatousJobanek) [inline test/e2e/parallel/space_test.go:503]: it's just a detail, but this format is easier to read
```suggestion
		subSpace := CreateSubSpace(t, awaitilities, 
			testspace.WithSpecParentSpace(parentSpace.Name), 
			testspace.WithTierName("appstudio"), 
			testspace.WithSpecTargetCluster(memberAwait.ClusterName), 
			testspace.WithDisableInheritance(true))
```

File: test/e2e/parallel/space_test.go [hunk 3]
```diff
@@ -480,6 +480,39 @@ func TestSubSpaces(t *testing.T) {
...
+		testsupportsb.CreateSpaceBinding(t, awaitilities.Host(), mur, subSpace, "viewer")
```
Comment (MatousJobanek) [inline test/e2e/parallel/space_test.go:504]: This technically creates a new SpaceBinding for the sub-space and for the same user that has access to the parent space. In other words, the test will pass no matter if the inheritance is disabled or not.

What you could do instead is to provision a brand new user and share the sub-space with this user
```suggestion
		_, secondMur := NewSignupRequest(awaitilities).
			ManuallyApprove().
			EnsureMUR().
			RequireConditions(ConditionSet(Default(), ApprovedByAdmin())...).
			Execute(t).Resources()
		testsupportsb.CreateSpaceBinding(t, awaitilities.Host(), secondMur, subSpace, "contributor")
```

apart from that, there is no "viewer" role in the appstudio NSTemplateTier - let's use "contributor" for example

File: test/e2e/parallel/space_test.go [hunk 4]
```diff
@@ -480,6 +480,39 @@ func TestSubSpaces(t *testing.T) {
...
+
+		// wait until subSpace has been provisioned
+		VerifyResourcesProvisionedForSpace(t, awaitilities, subSpace.Name)
+		// check that username and role from parentSpace was not inherited in the subSpace NSTemplateSet
+		spaceBindings, err := hostAwait.ListSpaceBindings(subSpace.Name)
+		require.NoError(t, err)
+		require.Equal(t,len(spaceBindings),1)
+		require.Equal(t,spaceBindings[0].Spec.SpaceRole, "viewer")
```
Comment (MatousJobanek) [inline test/e2e/parallel/space_test.go:512]: the SpaceBinding inheritance is done "virtually" - we don't replicate the SpaceBindings from the parents. What we do is that Space controller collects all SpaceBindings and updates the NSTemplateSet accordingly. That means that the number of the SpaceBinding CRs will be always the same no matter if the inheritance is disabled or not - what is different is the content of the NSTemplateSet.
```suggestion
		appstudioTier, err := hostAwait.WaitForNSTemplateTier(t, "appstudio")
		require.NoError(t, err)
		memberAwait.WaitForNSTmplSet(t, subSpace.Name, UntilNSTemplateSetHasSpaceRoles(
			SpaceRole(appstudioTier.Spec.SpaceRoles["contributor"].TemplateRef, secondMur.Name)))
```

File: testsupport/wait/member.go [hunk 5]
```diff
@@ -608,6 +608,19 @@ func UntilNSTemplateSetHasSpaceRoles(expected ...toolchainv1alpha1.NSTemplateSet
 	}
 }
 
+// UntilNSTemplateSetHasNoSpaceRoles returns a `NSTemplateSetWaitCriterion` which checks that the given
+// NSTemlateSet has no space roles.
+func UntilNSTemplateSetHasNoSpaceRoles() NSTemplateSetWaitCriterion {
```
Comment (MatousJobanek) [inline testsupport/wait/member.go:613]: this doesn't seem to be used

File: test/e2e/parallel/space_test.go (hunk 4 above)
Comment (mmulholla) [reply test/e2e/parallel/space_test.go]: I have modified the test and pushed my changes - please ignore the extra logs, will remove when all working. I seem to get the same roles in the NStemplate regardless of whether disable inheritance is set or not and not sure how to confirm I am using it. this in the NS template:
```
        spec:
          clusterResources:
            templateRef: appstudio-clusterresources-f08f06a-f08f06a
          namespaces:
          - templateRef: appstudio-tenant-76b9d37-76b9d37
          spaceRoles:
          - templateRef: appstudio-admin-85ef82d-85ef82d
            usernames:
            - ee86b9fd-e565-4128-a
          - templateRef: appstudio-contributor-004f85a-004f85a
            usernames:
            - 4a5388d8-3a96-4d7c-8
          tierName: appstudio
```
was thinking with disable inhertance I would just get contributor.  Also test code assumes only one role so if I add a second subspace I will need to allow for 2.

**Comment by mmulholla** [comment]
> Looks good, nice job!
> 
> I have few general questions:
> 
> * should we create subSpace with no inheritance using SpaceRequests, so that we verify if the disableInheritance parameter is being propagated correctly from SpaceRequest to Space objects. Or this will be a separate test?
> * should we create bigger hierarchy ( 3 spaces at least )
> 
> WDYT?

Space requests test will be later, when I add that support for host operator. Code is done just need to get the sapce code merged first.

I will create a bigger hierarchy once I get tte one subspace case working.

File: testsupport/wait/member.go (hunk 5 above)
Comment (mmulholla) [reply testsupport/wait/member.go:613]: Will delete if not needed but may need it as test evolves as one check is we don't get a parent space role if the sub space has disabled inheritance.

File: test/e2e/parallel/space_test.go (hunk 3 above)
Comment (mmulholla) [reply test/e2e/parallel/space_test.go]: done

File: test/e2e/parallel/space_test.go (hunk 1 above)
Comment (mmulholla) [reply test/e2e/parallel/space_test.go]: yes, adding a role for  the sub sapce.

File: test/e2e/parallel/space_test.go (hunk 4 above)
Comment (MatousJobanek) [reply test/e2e/parallel/space_test.go]: the code seems to be OK, no idea why it keeps inheriting the SpaceBinding from the parent space

Comment (MatousJobanek) [reply test/e2e/parallel/space_test.go]: It looks like that the tests are running fine, so it was maybe using some old version of the toolchain-common dependency.

**Review by MatousJobanek — APPROVED** [review]
Looks good :rocket: 
could you please drop the unused function `UntilNSTemplateSetHasNoSpaceRoles` and fix the linter issue? 
Otherwise, feel free to merge the toolchain-common repo and update the dependencies in these PRs :+1: :1st_place_medal:

File: testsupport/wait/member.go (hunk 5 above)
Comment (mmulholla) [reply testsupport/wait/member.go]: Removed it

**Comment by mmulholla** [comment]
> Looks good 🚀 could you please drop the unused function `UntilNSTemplateSetHasNoSpaceRoles` and fix the linter issue? Otherwise, feel free to merge the toolchain-common repo and update the dependencies in these PRs 👍 🥇

Have removed the function.

**Comment by mmulholla** [comment]
Updated to use latest toolchain-common from master, so ready for final review.

**Review by alexeykazakov — APPROVED** [review]

**Pull request merged**